Dynamic quorum size

hewison-chris opened the pull request:
The quorum size will be determined based on the number of active validators and no max will be specified.

hewison-chris committed:
Trivial missing whitespace

hewison-chris committed:
Dynamic SCP quorum size

`N = 3 x f + 1` where N is nodes and f is max faulty or byzantine
'S = 2 x f + 1' where S is the quorum size
To help ensure that quorum intersection occurs the number of nodes in a
quorum is calculated depending on the total number of active validators.

hewison-chris commented:
All green :-)

hewison-chris commented:
🟢

linked0 commented on source/agora/test/TimeDrift.d:
```diff
@@ -37,14 +36,6 @@ unittest
     // Make sure nodes have revealed their preimage for height 1
     network.waitForPreimages(GenesisBlock.header.enrollments, Height(1));
 
-    // sanity check for the generated quorum config
-    nodes.enumerate.each!((idx, node) =>
-        retryFor(node.getQuorumConfig().threshold == conf.consensus.max_quorum_nodes &&
-                node.getQuorumConfig().nodes.length == conf.consensus.max_quorum_nodes,
-            5.seconds,
-            format("Node #%s has invalid quorum config for test: %s",
-                idx, node.getQuorumConfig())));
-
```
Instead of removing this, can't we use the `quorumSize` in the code?  But if you think removing the code would be better, I have no objection.

hewison-chris replied:
I think now that the quorum size is calculated then this check is a bit redundant.

linked0 reviewed (approved):
LGTM

hewison-chris committed:
Trivial: use coins in place of coin units

hewison-chris committed:
Trivial: add some test documentation

hewison-chris committed:
Add unit test to print timings of quorum checks

It prints out the total number of nodes, the quorum size and
threshold for a slice as well as the time it takes to check the SCP
quorums have intersection to be safe based on the SCP verifier.
Currently it is set to check from 1 to 32 node counts which takes about
5 seconds on a MacBook.

hewison-chris commented:
Finally green again
